### PR TITLE
hyperbahn client first pass

### LINKED ISSUE #77: InitRequestInitiator should send current host:port if server is listening
So, this bit us because Hyperbahn uses the host:port value set in the original InitRequest message instead a value passed in as part of a Advertise request to determine where to reach the service. Needless to say I'll be filing a bug on that, but, as that issue is unlikely to get any traction, we'll need to patch our implementation ( and, in doing so, get up to spec with the protocol ).

The implications here are that ChannelInitializers for outbounds e.g. client requests will need to modify their ChannelPipeline to pass in the listening port of their owning TChannel instance. This is tricker than it sound due to the order of operations of how Pipelines are created ( typically pipelines are created prior to  listening but we'll have to modify client pipelines after a server pipeline is created )

### PR DESCRIPTION
**Pull request by willsalz**
What
====
Really basic Hyperbahn Client. Not yet 100% functioning due to TChannel bug #77 

Why
====
because hyperbahn. 

Misc
====
- Actually set the TTL on an outbound, oops.
- return an empty header map if Gson returns null on parse ( that would happen if there were no headers )
- Add a basic hyperbahn example
- fix some aggressive renaming of example request handlers
- implement basic hyperbahn client

Todo
====
- Fix #77 
- Perform a heartbeat with the Hyperbahn routers so we don't get disconnected
- implement the exponential backoff strategy for connecting

cc @truncs @brandoniles @junchaowu

### CODE REVIEW

File: tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java [hunk 1]
```diff
@@ -47,7 +48,8 @@ public String decodeEndpoint(ByteBuf arg1) {
     public Map<String, String> decodeHeaders(ByteBuf arg2) {
         String headerJSON = arg2.toString(CharsetUtil.UTF_8);
         arg2.release();
-        return new Gson().fromJson(headerJSON, HEADER_TYPE);
+        Map<String, String> headers = new Gson().fromJson(headerJSON, HEADER_TYPE);
```
Comment (truncs) [inline tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java]: this looks a bit weird, won't the headerJSON be null ? or its just {} ?

File: tchannel-example/src/main/java/com/uber/tchannel/hyperbahn/HyperbahnExample.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public static void main(String[] args) throws Exception {
+        TChannel tchannel = new TChannel.Builder("hyperbahn-example")
+                .register("ping", new PingRequestHandler())
+                .build();
+
+        tchannel.listen();
+
+        HyperbahnClient hyperbahn = new HyperbahnClient(tchannel);
+
+        Response<AdvertiseResponse> response = hyperbahn.advertise(tchannel.getServiceName(), 0);
```
Comment (truncs) [inline tchannel-example/src/main/java/com/uber/tchannel/hyperbahn/HyperbahnExample.java]: Ideally the end developer should really care about the response coming back from the advertise message ?

File: tchannel-example/src/main/java/com/uber/tchannel/json/JsonRequestHandler.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
 import com.uber.tchannel.api.Response;
 import com.uber.tchannel.api.ResponseCode;
 
-public class JsonDefaultRequestHandler extends DefaultRequestHandler<RequestPojo, ResponsePojo> {
+public class JsonRequestHandler extends DefaultRequestHandler<RequestPojo, ResponsePojo> {
```
Comment (truncs) [inline tchannel-example/src/main/java/com/uber/tchannel/json/JsonRequestHandler.java]: you are going to break my merge

File: tchannel-example/src/main/java/com/uber/tchannel/thrift/GetValueRequestHandler.java [hunk 4]
```diff
@@ -31,11 +31,11 @@
 
 import java.util.Map;
 
-public class GetValueHandlerDefault extends DefaultRequestHandler<KeyValue.getValue_args, KeyValue.getValue_result> {
+public class GetValueRequestHandler extends DefaultRequestHandler<KeyValue.getValue_args, KeyValue.getValue_result> {
```
Comment (truncs) [inline tchannel-example/src/main/java/com/uber/tchannel/thrift/GetValueRequestHandler.java]: ditto, about breaking the merge, we can move the renaming changes in a different PR ?

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Pattern;
+
+public class HyperbahnClient {
+
+    private static final String HOSTS_FILE_PATH = "/etc/uber/hyperbahn/hosts.json";
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: this makes little sense for OSS, internal developers should pass the file path manually even if it means some boilerplate code.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    private final TChannel tchannel;
+    private final Logger logger = LoggerFactory.getLogger(HyperbahnClient.class);
+    private List<InetSocketAddress> routers;
+
+    public HyperbahnClient(TChannel tchannel) throws IOException {
+        this.tchannel = tchannel;
+        this.routers = HyperbahnClient.loadRouters();
+    }
+
+    private static List<InetSocketAddress> loadRouters() throws IOException {
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: this function is weird, its a static function with sideeffects to routers.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            int port = Integer.parseInt(hostPortPair[1]);
+            InetSocketAddress router = new InetSocketAddress(host, port);
+            routers.add(router);
+        }
+
+        return routers;
+
+    }
+
+    public Response<AdvertiseResponse> advertise(
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: Advertise needs to be a continuous thing, I guess you are planning to add that later ?

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        final InetSocketAddress router = this.routers.get(new Random().nextInt(this.routers.size()));
+
+        Future<Response<AdvertiseResponse>> responseFuture = this.tchannel.callJSON(
+                router.getAddress(),
+                router.getPort(),
+                request,
+                AdvertiseResponse.class
+        );
+
+        return responseFuture.get(2, TimeUnit.SECONDS);
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: this makes the advertise message blocking.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+    }
+
+    public void stopAdvertising() {
+
+    }
+
+    public void shutdown() throws InterruptedException {
+        this.logger.info("Shutting down HyperbahnClient and TChannel.");
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: nitpick: just logger.info ?

File: tchannel-hyperbahn/src/test/java/com/uber/tchannel/hyperbahn/api/HyperbahnClientTest.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ * THE SOFTWARE.
+ */
+
+package com.uber.tchannel.hyperbahn.api;
+
+import com.uber.tchannel.api.TChannel;
+
+public class HyperbahnClientTest {
+
+    @org.junit.Test
```
Comment (truncs) [inline tchannel-hyperbahn/src/test/java/com/uber/tchannel/hyperbahn/api/HyperbahnClientTest.java]: why not import org.junit.Test
and then just use 
@Test

File: tchannel-hyperbahn/src/test/java/com/uber/tchannel/hyperbahn/api/HyperbahnClientTest.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    public void testAdvertise() throws Exception {
+        TChannel tchannel = new TChannel.Builder("hyperbahn-service").build();
+
+        HyperbahnClient hyperbahnClient = new HyperbahnClient(tchannel);
```
Comment (truncs) [inline tchannel-hyperbahn/src/test/java/com/uber/tchannel/hyperbahn/api/HyperbahnClientTest.java]: WIP ?

File: tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java (hunk 1 above)
Comment (willsalz) [inline tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java]: It can be a variety of things, and it depends on the caller. This prevents a NPE when Gson can't decode invalid json e.g. empty strings or nulls.

File: tchannel-example/src/main/java/com/uber/tchannel/json/JsonRequestHandler.java (hunk 3 above)
Comment (willsalz) [inline tchannel-example/src/main/java/com/uber/tchannel/json/JsonRequestHandler.java]: shouldn't have renamed all the example handlers then :)

Comment (willsalz) [inline tchannel-example/src/main/java/com/uber/tchannel/json/JsonRequestHandler.java]: but yeah can rename in a different PR.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 5 above)
Comment (willsalz) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: yup makes sense.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 7 above)
Comment (willsalz) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: yup, this is just a WIP

**Comment by willsalz** [comment]
Going to break this in to a few diffs, one for bug fixes, one for renaming, one for the actual hyperbahn client.
